File: server/mc/1.12.2-cleanroom/src/main/java/calebxzhou/rdi/mc/server/rcmd/RcmdServerSource112.java
package calebxzhou.rdi.mc.server.rcmd;

import calebxzhou.rdi.mc.rcmd.RcmdSource;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.util.text.TextComponentString;

import java.util.UUID;

public final class RcmdServerSource112 implements RcmdSource {
    private final EntityPlayerMP player;

    public RcmdServerSource112(EntityPlayerMP player) {
        this.player = player;
    }

    public EntityPlayerMP getPlayer() {
        return player;
    }

    @Override
    public String name() {
        return player.getGameProfile().getName();
    }

    @Override
    public UUID playerId() {
        return player.getUniqueID();
    }

    @Override
    public boolean hasPermission(String permission) {
        return player.server.getPlayerList().canSendCommands(player.getGameProfile());
    }

    @Override
    public void sendFeedback(String message) {
        player.sendMessage(new TextComponentString(message));
    }

    @Override
    public void sendError(String message) {
        player.sendMessage(new TextComponentString("[rcmd] " + message));
    }
}

File: server/mc/1.12.2-cleanroom/src/main/kotlin/calebxzhou/rdi/mc/server/rcmd/RcmdServerSource112.kt
package calebxzhou.rdi.mc.server.rcmd

import calebxzhou.rdi.mc.rcmd.RcmdSource
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.util.text.TextComponentString
import java.util.UUID

class RcmdServerSource112(val player: EntityPlayerMP) : RcmdSource {
    override fun name(): String = player.gameProfile.name

    override fun playerId(): UUID = player.uniqueID

    override fun hasPermission(permission: String): Boolean =
        player.server.playerList.canSendCommands(player.gameProfile)

    override fun sendFeedback(message: String) {
        player.sendMessage(TextComponentString(message))
    }

    override fun sendError(message: String) {
        player.sendMessage(TextComponentString("[rcmd] $message"))
    }
}
